python/tank/descriptor/bundle_cache_usage/logger.py

```diff
@@ -0,0 +1,281 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import threading
+import Queue
+
+from ...import LogManager
+from .errors import BundleCacheUsageTimeoutError
+from .database import BundleCacheUsageDatabase as BundleCacheUsageDatabase
+from . import BundleCacheUsageMyLogger as tmp_log
+
+
+log = LogManager.get_logger(__name__)
+
+
+class BundleCacheUsageLogger(threading.Thread):
+
+    """
+    Update bundle cache usage in a simple single table database.
+    To minimize performance impact, usage is logged using a
+    fire-and-forget scheme to a worker thread.
+
+    .. References::
+
+        https://docs.python.org/2/library/queue.html
+        https://docs.python.org/2/library/threading.html
+    """
+
+    DEFAULT_STOP_TIMEOUT = 10 # in seconds
+
+    KEY_BUNDLE_COUNT = "bundle_count"
+    KEY_LAST_USAGE_DATE = "last_usage_date"
+    KEY_USAGE_COUNT = "usage_count"
+
+    # keeps track of the single instance of the class
+    __singleton_lock = threading.Lock()
+    __singleton_instance = None
+
+    def __new__(cls, *args, **kwargs):
+        #
+        # Based on tornado.ioloop.IOLoop.instance() approach.
+        #
+        # See:
+        #   https://github.com/facebook/tornado
+        #   https://gist.github.com/werediver/4396488
+        #   https://en.wikipedia.org/wiki/Double-checked_locking
+        #
+        if not cls.__singleton_instance:
+            tmp_log.debug_worker_threading("__new__")
+            with cls.__singleton_lock:
+                if not cls.__singleton_instance:
+                    cls.__singleton_instance = super(BundleCacheUsageLogger, cls).__new__(cls, *args, **kwargs)
+                    cls.__singleton_instance.__initialized = False
+
+        return cls.__singleton_instance
+
+    def __init__(self, bundle_cache_root):
```

Comment:
as per previous comments, and your remarks, feels right not to pass the `bundle_cache_root` argument but the class can keep this logic internally.